<?xml version="1.0"?>
<launch>

    <arg name="standalone" default="true"/>

    <!-- ARGUMENTS - OAKD -->
    <arg name="oakd_mode"      default="rgbd" doc="OAK-D Modes [rgbd, stereo(only for debugging)]"/>
    <arg name="camera_name"    default="oakd"/>
    <arg name="lrcheck"        default="true"  /> 
    <arg name="extended"       default="false" /> 
    <arg name="subpixel"       default="true" />
    <arg name="rectify"        default="true"  />
    <arg name="depth_aligned"  default="false"  />
    <arg name="stereo_fps"     default="15"/> <!--https://docs.luxonis.com/projects/api/en/latest/components/nodes/stereo_depth/#currently-configurable-blocks-->
    <arg name="confidence"     default="200"/>
    <arg name="LRcheckthresh"  default="5"/>
    
    <group if="$(arg standalone)">
        <!-- INCLUDE DESCRIPTION LAUNCH FILE -->
        <include file="$(find akros_description)/launch/akros_description.launch" />
    </group>
    
    <!-- NODELET MANAGER -->
    <node pkg="nodelet" type="nodelet" name="nodelet_manager"  args="manager" output="screen"/>
    
    <!-- OAK-D NODELET -->
    <node name="oakd" pkg="nodelet" type="nodelet" output="screen" required="true"
        args="load akros_test/RGBDInertialNodelet nodelet_manager">
        <param name="camera_name" value="$(arg camera_name)"/>
        <param name="mode" value="$(arg oakd_mode)"/>
        <param name="lrcheck" value="$(arg lrcheck)"/>
        <param name="extended" value="$(arg extended)"/>
        <param name="subpixel" value="$(arg subpixel)"/>
        <param name="rectify" value="$(arg rectify)" />
        <param name="depth_aligned" value="$(arg depth_aligned)" />
        <param name="stereo_fps" value="$(arg stereo_fps)" />
        <param name="confidence" value="$(arg confidence)"/>
        <param name="LRcheckthresh" value="$(arg LRcheckthresh)"/>
    </node>   

    <!-- POINT_CLOUD_XYZ NODELET (DEPTH_IMAGE_PROC) -->
    <node pkg="nodelet" type="nodelet" name="pointcloud_xyz"
        args="load depth_image_proc/point_cloud_xyz nodelet_manager">
        <remap from="camera_info" to="/$(arg camera_name)/color/camera_info"/>
        <remap from="image_rect" to="/$(arg camera_name)/stereo/depth"/>    
        <remap from="points" to="/$(arg camera_name)/stereo/points"/>
    </node>
        
    <!-- VOXEL_GRID NODELET (PCL) -->
    <node pkg="nodelet" type="nodelet" name="voxel_grid" 
        args="load pcl/VoxelGrid nodelet_manager" output="screen">
        <remap from="~input" to="/$(arg camera_name)/stereo/points" />
        <remap from="~output" to="/$(arg camera_name)/stereo/points_vox" />
        <rosparam>
          filter_field_name: z
          filter_limit_min: 0.01
          filter_limit_max: 8
          filter_limit_negative: False
          leaf_size: 0.01
        </rosparam>
    </node>

</launch>